RickAndMortyFunApp/Screens/Episodes/EpisodeViewController.swift

```diff
@@ -10,33 +10,135 @@ import UIKit
 
 class EpisodeViewController: UIViewController {
 
-    private let cellId = "CellId"
-    var tableView = UITableView()
+    // MARK: - Views
+
+    var episodeTableView = UITableView()
+    var refreshControll = UIRefreshControl()
+
+    // MARK: - Properties
+    var itemsGrouped: [[Episode]] = []
+    var viewModel: ViewModel<ItemsResponse<Episode>>
+
+    // MARK: - Init
+
+    init(viewModel: ViewModel<ItemsResponse<Episode>> = .init()) {
+        self.viewModel = viewModel
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Life cycle
 
     override func viewDidLoad() {
         super.viewDidLoad()
         setupTableView()
+        setupRefreshControll()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        loadPage(1)
+
+        viewModel.stateUpdated = { [weak self] state in
+            self?.makeSectionsAndRowsArray()
+            self?.setFooterView(for: state)
+            self?.episodeTableView.reloadData()
+        }
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        viewModel.stateUpdated = nil
+    }
+
+    // MARK: - Actions
+
+    @objc private func refreshTableView() {
+        loadPage(1)
+        refreshControll.endRefreshing()
+    }
+
+    // MARK: - Private
+
+    private func loadPage(_ page: Int) {
+        let request = RickAndMortyRouter.getEpisode(page: page)
+        viewModel.load(request: request, page: page)
     }
 
-    fileprivate func setupTableView() {
-        view.addSubview(tableView)
-        tableView.fillSuperview()
-        tableView.dataSource = self
-        tableView.register(UITableViewCell.self, forCellReuseIdentifier: cellId)
-        tableView.tableFooterView = UITableViewHeaderFooterView()
+    private func makeSectionsAndRowsArray() {
+        itemsGrouped = []
```

Comment:
This could be rewritten as:
```
private func makeSectionsAndRowsArray() {
    itemsGrouped = Dictionary(grouping: viewModel.items) { $0.season }
        .sorted { $0.key < $1.key }
        .map { $0.value }
}
```